orm/src/main/java/org/hibernate/search/v6poc/entity/orm/model/impl/HibernateOrmPathFilterFactory.java

```diff
@@ -0,0 +1,290 @@
+ *         (i.e. a {@link org.hibernate.mapping.Collection}, which also encompasses maps)
+ *         containing only embeddeds (i.e. {@link Component})
+ *         or "atomic" values (i.e. {@link SimpleValue} but not its subtypes)
+ *         or "association" values (i.e. {@link org.hibernate.mapping.OneToMany},
+ *         {@link org.hibernate.mapping.ToOne} or {@link org.hibernate.mapping.Any})
+ *         then the path will be represented as either the "property path" of that prefix
+ *         and alternatively as the collection role.
+ *     </li>
+ *     <li>
+ *         Otherwise, if a prefix of the path that does not contain any multi-valued {@link Value}
+ *         can be resolved to a Hibernate ORM {@link Value}
+ *         that will be reported by ORM as dirty in a {@link PostUpdateEvent}
+ *         and is not an association
+ *         (i.e. {@link SimpleValue} but not its subtypes),
+ *         then the path will be represented as the "property path" of that prefix.
+ *
+ *         Rationale: such values are considered as atomic when persisting,
+ *         i.e. they will be reported by ORM as dirty in a {@link PostUpdateEvent}
+ *         whenever any of their parts changed.
+ *     </li>
+ *     <li>
+ *         Otherwise, if a prefix of the path can be resolved to a multi-valued {@link Value}
+ *         (i.e. a {@link org.hibernate.mapping.Collection}, which also encompasses maps)
+ *         containing only embeddeds (i.e. {@link Component})
+ *         or "atomic" values (i.e. {@link SimpleValue} but not its subtypes),
+ *         then the path will be represented as either the "property path" of that prefix
+ *         and alternatively as the collection role.
+ *
+ *         Rationale: such values are considered as atomic when persisting,
+ *         i.e. they will be reported by ORM as dirty in an {@link org.hibernate.event.spi.AbstractCollectionEvent}
+ *         whenever any of their parts changed.
+ *     </li>
+ *     <li>
+ *         Otherwise, no string representation can be assigned and an exception will be thrown.
+ *         This includes in particular all cases when the path points to a {@link javax.persistence.Transient} property,
+ *         or when a custom or {@link java.util.Optional} value extractor
+ *         is used before we can detect a prefix matching the conditions described above.
+ *     </li>
+ * </ul>
+ */
+public class HibernateOrmPathFilterFactory implements PojoPathFilterFactory<Set<String>> {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private final PersistentClass persistentClass;
+
+	public HibernateOrmPathFilterFactory(PersistentClass persistentClass) {
+		this.persistentClass = persistentClass;
+	}
+
+	@Override
+	public PojoPathFilter<Set<String>> create(Set<PojoModelPathValueNode> paths) {
+		// Use a LinkedHashSet for deterministic iteration
+		Set<String> pathsAsStrings = new LinkedHashSet<>();
```

Comment:
I doubt it will change anything since most of the time it will get resized when we add a double representation for one path, leading to a result pretty much identical to an uninitialized size. But ok, done.